src/dmap_import/util_api.py

```diff
@@ -47,23 +47,31 @@ def download_from_url(url: str, local_path: str) -> Optional[str]:
     download_log.log_start()
 
     max_retries = 3
-    for retry_count in range(max_retries):
+    for retry_count in range(max_retries + 1):
         download_log.add_metadata(retry_count=retry_count)
-        with requests.get(url, stream=True, timeout=60 * 5) as response:
-            # retry if status_code not 200
-            if response.status_code != 200:
-                if retry_count == max_retries - 1:
-                    response.raise_for_status()
-                # try again in 15 seconds
-                time.sleep(15)
-                continue
+        try:
+            response = requests.get(url, stream=True, timeout=15)
+            response.raise_for_status()
 
             # download file
             with open(local_path, "wb") as local_file:
                 for file_chunk in response.iter_content(chunk_size=None):
                     # chunk_size=None will write chunks in whatever size they are received
                     local_file.write(file_chunk)
-        break
+
+            response.close()
+            break
+
+        except Exception as exception:
+            if retry_count == max_retries:
```

Comment:
suggestion: I think you can replace this with a `for: else:` block:

```python
for try in range(max_tries):
  try:
    # do the thing
    break
  except Exception as e:
    # handle exception
else:
  # no breaks means we never succeeded
```

Reply:
If you want to use the exception outside the loop, you'd need to set it as a variable beforehand:
```
last_failure = None
for ...:
  try:
  ...
  except Exception as e:
    last_failure = (response, e)
    time.sleep(15)
else:
  (response, exception) = last_failure
  ...
```

Reply:
The still throws a pylint message on `last_failure`:

E0601: Using variable 'last_failure' before assignment (used-before-assignment)

Reply:
This configuration throws mypy errors.

Either:
```
error: "None" object is not iterable  [misc]
            (last_response, last_exception) = last_failure
```
with `last_failure = None` 

OR 

```
src/dmap_import/util_api.py:66: error: Incompatible types in assignment (expression has type
"tuple[Response, Exception]", variable has type "tuple[None, None]")  [assignment]
                last_failure = (response, exception)
                               ^~~~~~~~~~~~~~~~~~~~~
src/dmap_import/util_api.py:73: error: "None" has no attribute "status_code"  [attr-defined]
                staus_code=last_response.status_code,
                           ^~~~~~~~~~~~~~~~~~~~~~~~~
src/dmap_import/util_api.py:74: error: "None" has no attribute "text"  [attr-defined]
                response=last_response.text,
                         ^~~~~~~~~~~~~~~~~~
src/dmap_import/util_api.py:76: error: Argument 1 to "log_failure" of "ProcessLogger" has
incompatible type "None"; expected "Exception"  [arg-type]
            download_log.log_failure(last_exception)
                                     ^~~~~~~~~~~~~~
src/dmap_import/util_api.py:77: error: Exception must be derived from BaseException  [misc]
            raise last_exception
```
with `last_failure = (None, None)`

Reply:
Was able to bring this back with a new `exception` work flow, previous exceptions had the possibility of leaking API keys in clear text.

Reply:
its too bad that you can't do `for: elif:` blocks. 

i think if we wanted this to pass mypy you have to include an extra if block in side of the else block

```
last_failure = None
for ...:
    try:
        ... 
    except Exception as e:
        last_failure = (response, e)
        time.sleep(15)
else: 
    if last_failure:
        (response, exception) = last_failure
        ...
```

which is kind of clunky and i wouldn't be _that_ surprised if pylint yelled at you to use an `elif`.

i also think it reads ok as it is.